### ISSUE TITLE
Image Management

### ISSUE THREAD

**Issue by ElementalCrisis** [japanesemediamanager/jmmserver#427]
**Issue:** 

- Currently there is no image management system in JMM, images are downloaded and never checked again. Additionally, in JMM Desktop if the function Get Missing Data is executed for Related/Similar series, JMM will apply image settings to all series with missing data and download potentially hundreds to thousands of un-needed images. 

**Solution**

- A setting that compares local images to the ones currently available on AniDB, TvDB, Trakt and TMDb. Local images that do not match are deleted while site images with no corresponding local image are downloaded. 
- Images marked as favorite or default would be protected and never deleted unless no longer marked as favorite or default.
- Users wishing for more control can disable the global setting and run the per series setting similar to updating just a single show’s AniDB data. 
- A utility is added that shows all files marked for deletion, files are not deleted until reviewed by the user allowing complete control over image preference. 

**Enhancements**   
-	Only download the AniDB Poster for Related/Similar series when the function Get Missing Data is executed. 
-	Allows users to use their own custom images, uses can select images to be imported into JMM and displayed within their collection. These images would also be protected from being deleted since the user manually added them.
-	For Seiyuu’s, we do not need to download the character image for every single character they have ever voiced, we don’t even display these images anywhere. 

Related Issues:

- Duplicate Images For Series - https://github.com/japanesemediamanager/jmmclient/issues/454
- Images dont refresh on ui after downloading - https://github.com/japanesemediamanager/jmmclient/issues/366
- Word - https://github.com/japanesemediamanager/jmmclient/issues/454
- Word - https://github.com/japanesemediamanager/jmmclient/issues/454
- Word - https://github.com/japanesemediamanager/jmmclient/issues/454
- Word - https://github.com/japanesemediamanager/jmmclient/issues/454

**Comment by maxpiva** [comment]
This will help, and only keep the better resolution, > size.... Since
providers give us, like 40 images, and there is only 2 o 3 really
different, the other are the same image with different resolution / or
size...



http://www.codeproject.com/Articles/374386/Simple-image-comparison-in-NET

**Comment by bigretromike** [comment]
@maxpiva I think that @ElementalCrisis didnt want to compare same image in different resolution ?! (the issue dont say anything like that).

imo, we can do this without external libs, just compare size (file and resolution) and if any is different delete local, redownload.
The only downside (of any method here) will be that everytime you want to compare you will download it anyway - so it will be like downloading new set of images (ban alert). 

We could make it to:
1. read db, then read local file and if so compare file size - redownload if missing - (recheck, redownload, corrupt image fix)
2. I wouldn't go in road of duplicate images - we could compare images as @maxpiva propose but we can hit a wall sooner-or-later with upscaled crapy images..  Im not saying its bad or anything but it overdoing, we could just submit better images/poster to anidb and more people benefit from it ;-)

**Comment by ElementalCrisis** [comment]
AniDB only provides a single image and that's a poster. We don't really need to compare with them, it's with TvDB.

**Comment by da3dsoul** [comment]
Also the duplicate image check is good for another reason, and that is that we only download 10 (by default) images, and if they are 8 of the same image and 2 others, there's really not a good variety.

**Comment by ElementalCrisis** [comment]
@da3dsoul https://stackoverflow.com/questions/210650/validate-image-from-file-in-c-sharp

**Comment by bigretromike** [comment]
@ElementalCrisis you probably post wrong link, it have nothing related to this problem 👍

**Comment by ElementalCrisis** [comment]
Nope, right link. We were talking about this issue in Discord and talking about recent image corruption issues and @da3dsoul asked me to post a possible solution we found I'm this topic.

**Comment by bigretromike** [comment]
oh, because issue was about comapring images based on content not checking if image header is broken :-) move along then ;p

**Comment by ElementalCrisis** [comment]
Pasting @maxpiva message as well.

> Duplicate images, are not easy task, since you may have the same image at different resolutions, or the same image, with different kind of compression (png, jpg), or same resolution different compression, or same resolution, same compression but different metadata, because of that, for duplicate images, we need something usable. The propose is as follow, first deprecate per provider images, into, poster image, banner images, and background images, Second, using this or similar https://www.codeproject.com/Articles/374386/Simple-image-comparison-in-NET generate similar hashes from images,  so every series will download images from all providers, but only keep, the better resolution/compression one (can be configured).

**Comment by da3dsoul** [comment]
Image validation done

**Comment by Jimmyson** [comment]
Does this happen to include a Image Duplication remover?

**Comment by da3dsoul** [comment]
Nope, all it does so far is validate images when downloading them

**Comment by ElementalCrisis** [comment]
I've updated the Related Issues section, since this is one of the big items for 3.9 we don't need to have them all done but should focus on the critical issues and then work our way towards the improvements and new features.

I recall @RickDB having some experience in this field.
